### PR TITLE
Move timestamp bindings in STORAGE

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

This PR moves timestamp binding storage out of the coordinator and into the storage controller. The timestamp binding on-disk representation backed by a `Stash` instance. The intent is to replace `Stash` with a `persist` collection as soon as we are comfortable with that.

The timestamp binding feedback coming from workers is not a differential stream, it resembles more of an upsert stream where every statement invalidates the previous one. Previously we were storing the data in SQLite with the same semantics. In order to store in differential collection in `Stash` in a way that can play well with the standard collection compaction semantics the STORAGE controller differentializes all bindings received before storing them and "integrates" them before sending them down to workers.

There is some work left for the future of changing the workers to also work with a differential collection of bindings, just as if they received them from a `persist` collection directly, but that was too much of a change for a single PR.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

File: src/dataflow-types/src/client.rs [hunk 1]
```diff
@@ -441,6 +443,8 @@ impl<'a, C: Client + 'a> futures::stream::Stream for SelectStream<'a, C> {
             let cursor = self.cursor;
             let result = {
                 use futures::Future;
+                // TODO(petrosagg): this implementation seems incorrect as on every poll a new recv Future
+                // is constructed, polled once, and then dropped. Investigate
```
Comment (benesch) [inline src/dataflow-types/src/client.rs:447]: Eep!

File: src/dataflow-types/src/client/controller/storage.rs [hunk 2]
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
...
+                .collection::<PartitionId, ()>(&id.to_string())
+                .unwrap();
+
+            let collection_state = self.collection_mut(*id).expect("misisng_source_id");
+
+            let upper = ts_binding_stash.upper().unwrap();
+            let since = ts_binding_stash.since().unwrap();
+            let mut seal_frontier = upper.clone();
+
+            // Here we differentialize the bindings we got from workers
```
Comment (benesch) [inline src/dataflow-types/src/client/controller/storage.rs:324]: What would be really helpful for me, at least, is to describe this differentialization in prose! I've totally paged out the details of that conversation that we had long ago.

**Review by benesch — COMMENTED** [review]
Neat! This pretty much all makes sense to me. I think @frankmcsherry may have more opinions.

There's a lot of `unwraps()` around. What is our plan for those? I think in many cases we just need to exit if we can't persist timestamp bindings, but would be nice to handle those a bit more gracefully than panicking.

File: src/dataflow-types/src/client/controller.rs [hunk 3]
```diff
@@ -39,6 +40,11 @@ mod storage;
 pub use compute::{ComputeController, ComputeControllerMut};
 mod compute;
 
+pub struct ControllerConfig<C> {
+    pub dataflow_client: C,
+    pub state_dir: PathBuf,
+}
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller.rs:46]: Doc comments please!

File: src/dataflow-types/src/client/controller/storage.rs [hunk 4]
```diff
@@ -105,35 +123,68 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
     /// be repeatedly downgraded with `allow_compaction()` to permit compaction.
     pub async fn create_sources(
         &mut self,
-        mut bindings: Vec<CreateSourceCommand<T>>,
+        mut bindings: Vec<(GlobalId, (crate::sources::SourceDesc, Antichain<T>))>,
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:126]: Is it easy to explain this change?

File: src/dataflow-types/src/client/controller/storage.rs [hunk 5]
```diff
@@ -169,23 +221,6 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
...
-        // Validate that the ids exist.
-        self.as_ref()
-            .validate_ids(frontiers.iter().map(|(id, _)| *id))?;
-
-        let policies = frontiers
-            .into_iter()
-            .map(|(id, frontier)| (id, ReadPolicy::ValidFrom(frontier)));
-        self.set_read_policy(policies.collect()).await;
-        Ok(())
-    }
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:188]: This is .. unrelated refactoring or ?

File: src/dataflow-types/src/client/controller/storage.rs [hunk 6]
```diff
@@ -252,9 +287,13 @@ impl<'a, T: Timestamp + Lattice> StorageController<'a, T> {
...
 }
 
-impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
-    /// Acquire a mutable reference to the collection state, should it exist.
-    pub(super) fn collection_mut(
+impl<'a, C: Client<T>, T> StorageControllerMut<'a, C, T>
+where
+    T: Timestamp + Lattice + TryInto<i64> + TryFrom<i64>,
+    <T as TryInto<i64>>::Error: std::fmt::Debug,
+    <T as TryFrom<i64>>::Error: std::fmt::Debug,
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:294]: These constraints seem to be everywhere, and it would be great to localize them instead! For example, having them only on the `persist_timestamp_bindings` method, or better yet not at all and having the stash trait link them up only at the constructor time.

File: src/dataflow-types/src/client/controller/storage.rs [hunk 7]
```diff
@@ -105,35 +123,68 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
...
             }
         }
+
+        let mut dataflow_commands = vec![];
+
         // Install collection state for each bound source.
-        for binding in bindings.iter() {
-            let collection = CollectionState::new(binding.desc.clone(), binding.since.clone());
-            self.storage.collections.insert(binding.id, collection);
+        for (id, (desc, since)) in bindings {
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:150]: With the vector of bindings now being returned, I think it is more clear that we aren't looking at `since` when we do the work here. Like, we push a bunch of binding information that is potentially irrelevant from `since`s point of view. I'm not sure it is harmful as I'm not certain what the vector communicates, but wanted to point it out! (I think previously we .. validated nothing).

File: src/dataflow-types/src/client/controller/storage.rs [hunk 8]
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
...
+
+            let collection_state = self.collection_mut(*id).expect("misisng_source_id");
+
+            let upper = ts_binding_stash.upper().unwrap();
+            let since = ts_binding_stash.since().unwrap();
+            let mut seal_frontier = upper.clone();
+
+            // Here we differentialize the bindings we got from workers
+            ts_binding_stash
+                .update_many(bindings.iter().cloned().flat_map(|(pid, ts, offset)| {
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:326]: I would either defensively sort `bindings` here, or put some ALLCAPS comments to the effect that it should be sorted by `ts`. It looks like that is necessary for correctess, but ignore if I'm wrong.

File: src/dataflow-types/src/client/controller/storage.rs [hunk 9]
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
...
+                    let prev_offset = collection_state
+                        .last_reported_ts_bindings
+                        .entry(pid.clone())
+                        .or_default();
+
+                    let ts = ts.try_into().unwrap();
+                    let update = ((pid, ()), ts, offset.offset - prev_offset.offset);
+
+                    if seal_frontier.less_than(&ts) {
+                        seal_frontier = Max1Antichain::Singular(ts);
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:336]: The whole `Max1Antichain` thing continues to confuse me. It looks like `Stash` made that call, but .. yeah.

File: src/dataflow-types/src/client/controller/compute.rs [hunk 10]
```diff
@@ -123,7 +123,12 @@ impl<'a, T: Timestamp + Lattice> ComputeController<'a, T> {
     }
 }
 
-impl<'a, C: Client<T>, T: Timestamp + Lattice> ComputeControllerMut<'a, C, T> {
+impl<'a, C: Client<T>, T> ComputeControllerMut<'a, C, T>
+where
+    T: Timestamp + Lattice + TryInto<i64> + TryFrom<i64>,
+    <T as TryInto<i64>>::Error: std::fmt::Debug,
+    <T as TryFrom<i64>>::Error: std::fmt::Debug,
+{
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/compute.rs:131]: Mentioned elsewhere, but these constraints seem like noise. I think if you put them on an `impl` block for the (I believe) one method that requires them, then no one else should need them.

**Review by frankmcsherry — APPROVED** [review]
Looks good, thanks. I have a few comments about adding some docs and comments, and moving / subtracting type constraints around, but morally this seems good.

File: src/dataflow-types/src/client/controller/storage.rs (hunk 9 above)
Comment (benesch) [reply src/dataflow-types/src/client/controller/storage.rs:336]: What would you expect to see here with a normal antichain? Can we add that same API to `Max1Antichain`?

File: src/dataflow-types/src/client/controller/storage.rs [hunk 11]
```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
...
+            if PartialOrder::less_than(&compaction_frontier, &since) {
+                compaction_frontier = since;
+            }
+            if PartialOrder::less_than(&seal_frontier, &compaction_frontier) {
+                compaction_frontier = seal_frontier.clone();
+            }
+
+            ts_binding_stash.seal(seal_frontier).unwrap();
+            ts_binding_stash.compact(compaction_frontier).unwrap();
+            ts_binding_stash.consolidate().unwrap();
```
Comment (benesch) [inline src/dataflow-types/src/client/controller/storage.rs:366]: Oh I meant to ask: do we need to do something smarter than consolidating the timestamp bindings on every update?

File: src/dataflow-types/src/client/controller/storage.rs (hunk 9 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller/storage.rs:336]: That's what would be clearest. It's possible that I don't have the `Stash` requirements top of mind, but to the extent that `Max1Antichain` is a perf optimization, meh. It seemed like if it was an "I can be converted to `{ number, NULL }`" then it seemed that it could be a wrapper around `Antichain` with a `T: TotalOrder` constraint.

Not sure, but my opinions on "roll your own" TD/DD stuff has evolved. :D No rush to change things that work, but I end up needing to re-understand if it does what I think when it is a new type to me.

Comment (benesch) [reply src/dataflow-types/src/client/controller/storage.rs:336]: I suppose it's sort of a perf optimization, but yeah, I'd phrase the requirement as restricting the antichain to a type we know we can represent in SQL. 

> then it could be a wrapper around Antichain with a T: TotalOrder constraint.

That sounds right! I think the Frank that reviewed the stash PR hadn't come up with that yet. :D

**Comment by guswynn** [comment]
Would it be reasonable to file a TODO task to remove the `blocking` io happening within timely works (and I suppose, the `coord` thread as well)

**Comment by benesch** [comment]
Where does the blocking on timely worker threads happen? It seems to only block the coord thread to me! The rule about blocking in an async task doesn't really apply here, since it has a dedicated thread, so not worried about that. It does impose a latency tax on the rest of the system, though, which is probably worth an issue.

File: src/expr/src/id.rs [hunk 12]
```diff
@@ -131,12 +131,37 @@ impl fmt::Display for SourceInstanceId {
...
 ///     None -> sources that have no notion of partitioning (e.g file sources)
-#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
 pub enum PartitionId {
     Kafka(i32),
     None,
 }
 
+use bytes::{Buf, BufMut};
+impl mz_persist_types::Codec for PartitionId {
```
Comment (aljoscha) [inline src/expr/src/id.rs:141]: All the other source struct for which we added `Codec` impls have protobuf-based impls. For example, take a look at `SourceTimestamp`.

**Comment by guswynn** [comment]
@benesch oh I see, `StorageControllerMut` methods are only called on the coordinator thread?
definitely we want a followup to resolve this latency bottleneck, the entire system could wait on sqlite io

also, will this pr need a migration to read the old bindings out of the catalog?

**Comment by benesch** [comment]
> @benesch oh I see, `StorageControllerMut` methods are only called on the coordinator thread? 

Yep!

> also, will this pr need a migration to read the old bindings out of the catalog?

Yeah, @petrosagg said he was working on them!

File: misc/python/materialize/mzcompose/services.py [hunk 13]
```diff
@@ -577,6 +578,11 @@ def __init__(
         if validate_catalog:
             entrypoint.append("--validate-catalog=/share/mzdata/catalog")
 
+        if validate_timestamp_bindings:
+            entrypoint.append(
+                "--validate-timestamp-bindings=/share/mzdata/catalog/controller/timestamp_bindings"
+            )
+
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply misc/python/materialize/mzcompose/services.py:530]: changed the flat to that

File: src/testdrive/src/action/verify_timestamp_compaction.rs [hunk 14]
```diff
@@ -70,7 +75,22 @@ impl Action for VerifyTimestampCompactionAction {
...
+                            .iter()
+                            .unwrap()
+                            .map(|((pid, _), ts, offset)| {
+                                (
+                                    pid,
+                                    ts.try_into().unwrap_or_else(|_| panic!()),
+                                    MzOffset { offset },
+                                )
+                            })
+                            .collect();
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/testdrive/src/action/verify_timestamp_compaction.rs:89]: yep, changed

**Comment by petrosagg** [comment]
Alright, this is finally ready for a final review. Some notes here for reviewers:

* @benesch all `.unwrap()`ed errors generated within Storage controller are all wired up to Result return types. Since the compute controller calls some of the now fallible methods all the transitive dependants are also changed to return a Result if they didn't already. The Result wrapping stops at the coordinator where all the calls to the controllers are `.unwrap()`ed. This is what we were doing already and I didn't want to delay this PR with further error type refactoring.
* @aljoscha the `Codec` implementation for `PartitionId` is now based on a protobuf schema
* @frankmcsherry I wasn't able to contain the extra trait bounds on the timestamp just to the methods that need it. Attempting to do so forced me to annotate every single method that transitively called the methods that lead to `Stash` calls and in the end there were more type annotations than the current version. If you have any ideas on how to reduce the noise I'm happy to try another approach.
* @benesch I'll like your eyes in particular on how the catalog migration is handled. Specifically, the migration code resolves this path `/mzdata/catalog/../controllers/storage/timestamp_bindings`. It seemed simpler than wiring up a separate top level path parameter just for this reason but if you think this is a smell I can wire up another parameter that is just the `/mzdata` path.

File: src/dataflow-types/src/client/controller/storage.rs (hunk 8 above)
Comment (petrosagg) [reply src/dataflow-types/src/client/controller/storage.rs]: no, stash doesn't require the operator to be sorted

**Comment by frankmcsherry** [comment]
> Attempting to do so forced me to annotate every single method that transitively called the methods that lead to Stash calls and in the end there were more type annotations than the current version.

I'm going to guess that those annotations were on the code that actually make the assumptions, though, vs the code that doesn't make the assumptions? We've been working on removing these constraints wherever possible, and having them creep back in because it makes other parts of the code easier is ... frustrating.

**Comment by petrosagg** [comment]
> I'm going to guess that those annotations were on the code that actually make the assumptions, though, vs the code that doesn't make the assumptions? We've been working on removing these constraints wherever possible, and having them creep back in because it makes other parts of the code easier is ... frustrating.

The code that actually makes the assumptions is quite a bit larger than I thought infecting multiple methods of the storage controller and multiple methods of the compute controller. Basically any function that can transitively cause a storage compaction needs to have these bounds in place, and it turns out that the majority of the functions have some control flow that leads there. It's not about making the code easier, it's less noise overall. If you feel strongly about this I can put it on a function by function basis and just live with these bounds all over the place

File: src/dataflow-types/src/client/controller/storage.rs [hunk 15]
```diff
@@ -326,11 +496,25 @@ impl<'a, T: Timestamp + Lattice> StorageControllerMut<'a, T> {
...
+                    .collection::<PartitionId, ()>(&id.to_string())?;
+
+                let mut since = ts_binding_stash.since()?;
+                since.extend(
+                    frontier
+                        .iter()
+                        .map(|t| t.clone().try_into().expect("timestamp overflowed i64")),
+                );
+                ts_binding_stash.compact(since.borrow())?;
+                ts_binding_stash.consolidate()?;
```
Comment (benesch) [inline src/dataflow-types/src/client/controller/storage.rs:516]: Consolidating on every update might be expensive. Not sure! It matches what we currently do, perhaps, and so is maybe fine?

File: src/coord/src/catalog/storage.rs [hunk 16]
```diff
@@ -128,22 +148,80 @@ const MIGRATIONS: &[&str] = &[
...
+    // Migrates timestamp bindings from the coordinator's catalog to STORAGE's internal state
+    // Introduced in v0.25.0.
+    &|data_dir_path: &Path, tx: &rusqlite::Transaction| {
+        let source_ids = tx
+            .prepare("SELECT DISTINCT sid FROM timestamps")?
+            .query_and_then([], |row| Ok(row.get::<_, SqlVal<GlobalId>>(0)?.0))?
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        std::fs::create_dir_all(data_dir_path.join("controllers/storage/")).unwrap();
+        let stash = Stash::open(&data_dir_path.join("controllers/storage/timestamp_bindings"))
```
Comment (benesch) [inline src/coord/src/catalog/storage.rs:176]: I was thinking that we'd have one `Stash` per layer, and namespace collection IDs as necessary. So `controllers/storage` would have `timestamp-bindings-1`, `timestamp-bindings-2`, etc. That leaves open the possibility of adding transactions to `Stash`, and then allowing each layer to transactionally update any of its collections. (Adding cross-collection transactions seems straightforward; cross-`Stash` transactions are far fetched though.)

**Review by benesch — APPROVED** [review]
> The Result wrapping stops at the coordinator where all the calls to the controllers are .unwrap()ed. This is what we were doing already and I didn't want to delay this PR with further error type refactoring.

This is great, thanks!

> It seemed simpler than wiring up a separate top level path parameter just for this reason but if you think this is a smell I can wire up another parameter that is just the /mzdata path.

It seems like you already went this direction? I left you some comments about dropping the `mzdata/catalog` argument now, though, since it seems you're passing through `mzdata` and could just construct `mzdata/catalog` on the fly.

**Comment by petrosagg** [comment]
@benesch I addressed all your comments with the exception of consolidation. I'd like to see how that performs. One interesting thing to note is that this new implementation doesn't add new rows in the database as the upper ticks forwards because we're storing the bindings in differentialized form. Whereas the current main repeatedly stores the same statement over and over again so the data in SQLite should be significantly smaller for sources that don't produce data continuously

File: src/dataflow-types/src/client/controller/storage.rs [hunk 17]
```diff
@@ -267,11 +364,90 @@ impl<T: Timestamp + Lattice> StorageController for Controller<T> {
...
+            // (PartitionId::Kafka(2), 1, +10)
+            // (PartitionId::Kafka(1), 2, +2)  <-- This is how much the offset changed, 7 - 5 = 2
+            // (PartitionId::Kafka(2), 2, +7)
+            //
+            // This representation allows us to compact timestamp bindings simply by adding
+            // together offsets and collapsing their timestamps. For example, if we were to compact
+            // thorugh timestamp 2 the collection would contain the following updates:
+            //
+            // (PartitionId::Kafka(1), 1, +7)  <-- 5 + 2 = 7
+            // (PartitionId::Kafka(2), 1, +17) <-- 10 + 7 = 17
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client/controller/storage.rs:407]: Probably want the timestamps to be `2` here.

File: src/dataflow-types/src/client/controller/storage.rs [hunk 18]
```diff
@@ -267,11 +364,90 @@ impl<T: Timestamp + Lattice> StorageController for Controller<T> {
...
+            ts_binding_stash.update_many(bindings.iter().cloned().flat_map(
+                |(pid, ts, offset)| {
+                    let prev_offset = collection_state
+                        .last_reported_ts_bindings
+                        .entry(pid.clone())
+                        .or_default();
+
+                    let ts = ts.try_into().expect("timestamp overflowed i64");
+                    let update = ((pid, ()), ts, offset.offset - prev_offset.offset);
+
```
[the remark this answers is not in the document]
Comment (frankmcsherry) [reply src/dataflow-types/src/client/controller/storage.rs:417]: The `prev_offset` update below maybe addresses this, but .. is this list sorted? It's all a bit confused, having these state updates applied in a `flat_map` that is an argument to `update_many`.

File: src/stash/src/lib.rs [hunk 19]
```diff
@@ -117,26 +118,36 @@ impl Stash {
...
+                )?;
+                tx.execute(
+                    "INSERT INTO uppers (collection_id, upper) VALUES ($collection_id, $upper)",
+                    named_params! {"$collection_id": collection_id, "$upper": 0_i64},
+                )?;
+                collection_id
+            }
+        };
+
+        tx.commit().unwrap();
```
Comment (mjibson) [inline src/stash/src/lib.rs:150]: Why `unwrap` instead of `?`?

File: src/dataflow-types/src/client/controller/storage.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/dataflow-types/src/client/controller/storage.rs]: I updated this section to have an explicit sort step and the calculations in a for loop. Having multiple updates for `pid` at the same timestamp will just create zero diff changes, which is fine.

File: src/stash/src/lib.rs (hunk 19 above)
Comment (petrosagg) [reply src/stash/src/lib.rs]: oops. Debugging leftover, I was trying to hunt down an error with a stacktrace. fixed

**Comment by cjubb39** [comment]
Any chance that re-enabling the tests disabled in #10942 passes CI now? :)

From `test/testdrive/kafka-exactly-once-sink.td`
